File: src-backend/videoscribe/domain/ipc_models.py
from dataclasses import dataclass, field
from typing import Optional, Dict, Any, List

@dataclass
class StartPayload:
    video_path: str
    model: str = "medium"
    language: str = "auto"
    use_vad: bool = False
    use_batch: bool = True
    batch_size: int = 16

@dataclass
class IpcCommand:
    action: str
    job_id: Optional[str] = None
    payload: Optional[Dict[str, Any]] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'IpcCommand':
        return cls(
            action=data.get("action", "unknown"),
            job_id=data.get("job_id"),
            payload=data.get("payload")
        )

@dataclass
class JobStateData:
    job_id: str
    status: str
    progress: Optional[int] = None
    language: Optional[str] = None
    error_message: Optional[str] = None
    runtime_device: Optional[str] = None
    runtime_compute_type: Optional[str] = None

@dataclass
class CueData:
    id: str
    ordinal: int
    start_ms: int
    end_ms: int
    text: str

@dataclass
class SegmentBatchData:
    job_id: str
    cues: List[Dict[str, Any]] # simplified for now, or List[CueData]

@dataclass
class IpcEvent:
    event: str
    data: Dict[str, Any]
    version: int = 1

    def to_dict(self) -> Dict[str, Any]:
        return {
            "version": self.version,
            "event": self.event,
            "data": self.data
        }

File: src-tauri/src/domain/ipc_models.rs
use serde::{Deserialize, Serialize};
use crate::domain::stt_job::SttCue;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StartPayload {
    pub video_path: String,
    pub model: String,
    pub language: String,
    pub vad_engine: String,
    pub use_batch: bool,
    pub batch_size: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum WorkerCommand {
    Start {
        job_id: String,
        payload: StartPayload,
    },
    Cancel {
        job_id: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct JobStateData {
    pub job_id: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub progress: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime_device: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime_compute_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct SegmentBatchData {
    pub job_id: String,
    pub cues: Vec<SttCue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ErrorData {
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", content = "data", rename_all = "snake_case")]
pub enum WorkerEventData {
    JobState(JobStateData),
    SegmentBatch(SegmentBatchData),
    Error(ErrorData),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkerEvent {
    pub version: u32,
    #[serde(flatten)]
    pub data: WorkerEventData,
}
